### PR TITLE
add ImportC C compiler

### PR DESCRIPTION
**Pull request by WalterBright**
This adds an ANSI C11 C compiler to dmd so it can import and compile C files directly. See `changelog/ImportC.md` for details.

This is a prototype implementation, known shortcomings are in the changelog.

It doesn't alter D in any way.

### CODE REVIEW

**Comment by UplinkCoder** [comment]
This is a 3000 line pr... and preliminary support for parsing a related but different language  it's labeled trivial...

**Comment by WalterBright** [comment]
```
Undefined symbols for architecture x86_64:
  "Token::toChars(unsigned short)", referenced from:
      test_tokens() in cxxfrontend.o
  "Target::isVectorOpSupported(Type*, unsigned short, Type*)", referenced from:
      test_target() in cxxfrontend.o
```
This is a bit baffling. The declarations in the code are correct.

File: src/dmd/cparse.d
Comment (maxhaton) [inline src/dmd/cparse.d:4]: C98?

**Comment by maxhaton** [comment]
C:
```C
typedef float real;

typedef struct impl {
    void* x;
    char lazy;
} module;

void does(int x)
{
    module copy;
}
```
D:
```D
import cdecl;

extern(C)
int main()
{
    impl g;

    module x;
    pragma(msg, __traits(allMembers, impl)); //tuple("x", "lazy")
    return g.lazy;
}
```

**Comment by adamdruppe** [comment]
This appears completely useless to me. First, it doesn't "just work" since you need to run cpp over the file before you can use it, and since that expands macros, it must be done for each different target, meaning it must be part of the build system. This offers no advantage over the (not great) existing solutions and shares many of their problems including creating huge files.

Even if that just worked, C bindings tend to rely on the preprocessor to do common and necessary tasks like defining constants. Without that, you can't even reasonably use real world C headers. Possibly  a limited processor could translate them, but even that data is gone after cpp goes through it, so there's no real hope to recover it into a usable form.

I think trying to run the cpp over a whole D file like dpp does is problematic... but like *something* must be done. dstep's approach actually does a reasonably good job, at the cost of requiring a bit of hand editing and some setting up of other files too.

I do think there's some potential in integrating with dmd... but this PR as it is looks like more harm than good. And I'm not sure it is worth the time sink it will take to get it to a positive point.

**Comment by WalterBright** [comment]
@adamdruppe All good points, and I anticipated this reaction. But may I ask that the discussion of "should we do this at all" be moved to the newsgroup? I expect it to be long, and threading in the n.g. will make it easier to navigate. I already started the thread on it.

Please keep this one confined to a review of the PR. Thanks!

**Comment by Imperatorn** [comment]
I think this is an interesting prototype. I don't see it as useless. It needs some work ofc, but it could potentially become quite handy

**Comment by maxhaton** [comment]
```C
struct context;

struct context {
    float fps;
}; 
```
Fails to compile with `Error: no struct-declarator-list for struct context`, which in turn breaks every C header I've tried so far.

**Comment by WalterBright** [comment]
@maxhaton That one fails because the semantic processing is in the D part. I haven't addressed that yet.

**Comment by 12345swordy** [comment]
@WalterBright try to follow the "separations of concern" design principle when implementing the c compiler. Don't be afraid of creating new files for this.

**Comment by WalterBright** [comment]
Buildkite Style appears to have a problem:
```
/var/lib/buildkite-agent/builds/ci-agent-6af57101-f5f2-406d-935a-c6100716b586-8/dlang/dmd/dmd/src/dmd/cparse.d(219:18)[error]: Expected `identifier` instead of `typedef`
```
Looks like Style's list of tokens needs updating.

**Comment by WalterBright** [comment]
I have no idea why CyberShadow/DAutoTest is failing, the log file says it succeeded.

**Comment by thewilsonator** [comment]
for DAutoTest failures click `View` on `Build log    View` and then scroll _alllll_ the way down to the end, then scroll up a bit pas the stack trace of the build command, past the make invocation errors and the source of the problem is usually right there (except if you get LaTeX errors which result in a massive buildlog and are indecipherable).

**Comment by RazvanN7** [comment]
@WalterBright There is a hidden file in `src/` called `.dscanner.ini` where the rules for style checks are configured. You can disable checks for some modules there: https://github.com/dlang/dmd/blob/master/src/.dscanner.ini#L107 . I assume that  the `trust_too_much` variable might store the files for which checks are disabled.

**Comment by RazvanN7** [comment]
@WalterBright It seems that dscanner can exclude individual files from checks, so one way to go about this is to manually include this file in the disable check list of every check [1]

[1] https://github.com/dlang-community/D-Scanner/blob/2963358eb4a24064b0893493684d4075361297eb/src/dscanner/analysis/config.d#L225

**Comment by WalterBright** [comment]
That won't fix the problem, as it's Style's parser that's failing.

**Comment by TungstenHeart** [comment]
even if as you said D is not affected, I think that a `-preview=ImportC` is necessary.

File: src/dmd/cparse.d
Comment (dkorpel) [inline src/dmd/cparse.d:3088]: Add a slash to make it a DDoc comment

File: src/dmd/cparse.d
Comment (dkorpel) [inline src/dmd/cparse.d:3103]: Add a slash to make it a DDoc comment

File: src/dmd/cparse.d
Comment (TungstenHeart) [reply src/dmd/cparse.d:808]: mmmh there's not `goto` for this `Lerr`, just like @dkorpel  said.

File: changelog/ImportC.md
Comment (adamdruppe) [reply changelog/ImportC.md:90]: And how does it determine the C module name? Does it ever go into a package namespace? How does this interact with the import path?

File: src/dmd/parse.d
Comment (WalterBright) [reply src/dmd/parse.d:6930]: added

File: src/dmd/cparse.d
Comment (WalterBright) [reply src/dmd/cparse.d]: deleted

File: src/dmd/cparse.d
Comment (WalterBright) [reply src/dmd/cparse.d]: added message

File: src/dmd/cparse.d
Comment (WalterBright) [reply src/dmd/cparse.d]: done

**Comment by 12345swordy** [comment]
I don't have the ability to merge this though....

File: src/dmd/cparse.d
Comment (ibuclaw) [inline src/dmd/cparse.d:794]: Preferably we'd have unimplemented errors instead of an ICE in the parser.

**Comment by 12345swordy** [comment]
Should the d language foundation create an repository that focus on compiling popular c libraries? That is a fast way to find bugs.

File: changelog/ImportC.md
Comment (Luhrel) [reply changelog/ImportC.md:90]: 2. Do we have a `pragma` instruction to fix this ? something like `pragma(importFile, "wayland-util.h")` would be appreciated.

File: src/dmd/cparse.d
Comment (WalterBright) [reply src/dmd/cparse.d]: added message

**Comment by Imperatorn** [comment]
> Should the d language foundation create an repository that focus on compiling popular c libraries? That is a fast way to find bugs.

Good idea. When this gets in there should definitely be a list of libs to compile added to the test suite.

File: changelog/ImportC.md
Comment (ibuclaw) [reply changelog/ImportC.md:90]: > 2. Do we have a `pragma` instruction to fix this ? something like `pragma(importFile, "wayland-util.h")` would be appreciated.

I'm not thinking too hard about it, but there are features in the language that don't necessarily mean we need yet another feature added.

```
module c.wayland_util;
mixin(import("wayland-util.h"));
```

**Comment by ibuclaw** [comment]
I'm going to raise the usual red tape around this:

- Is there a bug report?
- When no, was a DIP written?
- Has this gone through community review?
- Has this been approved for inclusion in the language?

**Comment by WalterBright** [comment]
> Is there a bug report?

No. It's not a bug.

> When no, was a DIP written?

No. It does not change the D language.

> Has this gone through community review?

https://digitalmars.com/d/archives/digitalmars/D/Add_ImportC_compiler_to_dmd_349611.html

> Has this been approved for inclusion in the language?

It's not part of the D language. It's a new tool.

**Comment by Imperatorn** [comment]
> I'm going to raise the usual red tape around this:
> 
> * Is there a bug report?
> * When no, was a DIP written?
> * Has this gone through community review?
> * Has this been approved for inclusion in the language?

Will this change existing D in any way? 
Like, if this is added, will everything (not using ImportC) stay exactly the same?

If so, I guess you could allow it without a DIP?

Review will ofc be necessary, and is what we're doing now 😃

File: changelog/ImportC.md
Comment (Luhrel) [reply changelog/ImportC.md:90]: > > 1. Do we have a `pragma` instruction to fix this ? something like `pragma(importFile, "wayland-util.h")` would be appreciated.
> 
> I'm not thinking too hard about it, but there are features in the language that don't necessarily mean we need yet another feature added.
> 
> ```
> module c.wayland_util;
> mixin(import("wayland-util.h"));
> ```

Oh nice. However, `import("filename");` doesn't seem to be documented [here](https://dlang.org/spec/module.html#import-declaration).

**Comment by WalterBright** [comment]
> Will this change existing D in any way?

No.

> Like, if this is added, will everything (not using ImportC) stay exactly the same?

Yes.

> If so, I guess you could allow it without a DIP?

It needs a DIP about as much as dpp or dub does.

File: changelog/ImportC.md
Comment (BorisCarvajal) [reply changelog/ImportC.md:90]: It's documented [here](https://dlang.org/spec/expression.html#import_expressions).

**Comment by wilzbach** [comment]
> It needs a DIP about as much as dpp or dub does.

Those are third-party tools and improvements to dub need to go through the DEP process.


> Will this change existing D in any way?

No, as it will add another half-baked feature instead of focusing on things that _change existing D_.
For example, we haven't even finished nor officially released the C/C++ header generator :/

**Comment by Imperatorn** [comment]
> For example, we haven't even finished nor officially released the C/C++ header generator :/

How much work (guesstimation) do you think it would be to finish it?

**Comment by adamdruppe** [comment]
>>    Will this change existing D in any way?
> No.

```
and in the D source file:

     import stdio;  // reads stdio.c and compiles it
```

https://github.com/dlang/dmd/pull/12507#pullrequestreview-655115578


Both these statements cannot be simultaneously true.

**Comment by MoonlightSentinel** [comment]
> > For example, we haven't even finished nor officially released the C/C++ header generator :/
> 
> How much work (guesstimation) do you think it would be to finish it?

That's different though. A lot of improvements were implemented for `dtoh` since its introduction and there are only a few remaining issues AFAICT (I have patches/ideas for most of them but didn't get around to making proper PR's yet).

File: changelog/ImportC.md
Comment (MoonlightSentinel) [reply changelog/ImportC.md:90]: > 1. .d is looked for, then .c

Note that this only applies to one import path, not the entire list. The following code will be changed by this patch:

```
Project directory:

main/
  - app.d
  - stuff.c
util/
  - stuff.d
```

```d
// main/app.d

import stuff;

void main()
{
    import std.stdio : writeln;
    writeln(foo());
}
```

```c
// main/stuff.c
int foo()
{
    return 1;
}
```

```d
// util/stuff.d
int foo()
{
    return 2;
}
```

DMD:
```
dmd -Imain -Iutil -i main/app.d
```

The resulting binary will now print `1` instead of `2`.

---

THis project layout might be uncommon but at least warrants a warning in the changelog

File: src/dmd/dmodule.d
Comment (MoonlightSentinel) [inline src/dmd/dmodule.d:1060]: What about C header files? Code using external libraries might only have access to the header files, not the actual sources.

**Comment by adamdruppe** [comment]
I think I'd feel better about this if it didn't hack D's import syntax.

We have `#name` available in the D compiler. Maybe it should just go ahead and use that so you can specify the C filename - .c or .h, local or system path - and then also specify the D namespace you want it enter.

```
#import c_foo = "foo.h"
```

or something. then you specify the filename and D namespace it is limited to.

**Comment by 12345swordy** [comment]
@WalterBright  I am quite curious, why couldn't this be made into a plugin?

**Comment by WalterBright** [comment]
> Both these statements cannot be simultaneously true.

Picking up a .c file in the import path does not change the D language.

File: src/dmd/dmodule.d
Comment (WalterBright) [reply src/dmd/dmodule.d:1060]: One thing at a time.

**Comment by WalterBright** [comment]
> hack D's import syntax

There's no purpose to changing the syntax. The whole point is to *not care* what it's importing.

**Comment by WalterBright** [comment]
> why couldn't this be made into a plugin?

What problem would that solve?

**Comment by adamdruppe** [comment]
On Wed, May 12, 2021 at 04:52:15PM -0700, Walter Bright wrote:
> There's no purpose to changing the syntax. The whole point is to *not care* what it's importing.

I want to import the C file called "sys/foo.bar".

How do I do that?

**Comment by WalterBright** [comment]
> How do I do that?

Create foo.c file:
```
// My configuration #define's
#define FEATUREX TRUE
#define AUTHOR "adamdruppe"

#include "sys/foo.bar.h"
```
```
import foo;
```
Note that this also implicitly makes the `#include <file.h>` syntax workable without even trying and with *no* additions to DMD.

However, I fully expect that experience may suggest a better way. But currently we have no experience with this, and it would be a waste of time developing a complex method that may turn out to be worse. The existing method is plenty good enough to achieve liftoff.

**Comment by adamdruppe** [comment]
hmmm yeah, that's not a bad idea (if the preprocessor works!).

**Comment by 12345swordy** [comment]
> 
> 
> > why couldn't this be made into a plugin?
> 
> What problem would that solve?

It solves the problem of having to modifying the source code of dmd, if this is just a mere tool.

**Comment by Imperatorn** [comment]
@WalterBright some quick questions:

1. How capable is ImportC atm? Like, what can it currently by used for (if it was merged now)? 

2. What do you think about having a pre-processor built in, pros/cons?

3. What is your vision on how ImportC could/should be used? 

Thanks!

**Comment by WalterBright** [comment]
I've annotated shortcomings in the cparse.d file with the "TODO" string. Some are also in the changelog entry.

The biggest issue is I have deliberately refrained from altering the rest of dmd to accommodate C semantics where they diverge from D, in order to avoid over complicating this PR. Once this is merged, I can do PRs incrementally for those issues.

Points 2 and 3 are meta discussions, which should be discussed in the n.g., and indeed have been. In order to keep this PR discussion, which is not threaded, manageable, please use it only for a code review of the PR.

https://digitalmars.com/d/archives/digitalmars/D/Add_ImportC_compiler_to_dmd_349611.html

**Review by ibuclaw — COMMENTED** [review]
It's good that static assert is given a good stress, though the rest of the parser also needs to have tests for it as well.

Will try to find time to have a deep dive later.

**Comment by ibuclaw** [comment]
It is worrying that TargetC has not been expanded. This means that the C -> D mapping is going to be wrong.

**Comment by WalterBright** [comment]
> It is worrying that TargetC has not been expanded.

How so? Note that this PR avoids adding `if (compiling C)` blocks to D semantic routines because this PR is large enough already. And if this PR twists in the wind for months, it'll become very difficult to keep it rebased as the rest of dmd shifts around.

**Comment by ibuclaw** [comment]
> > It is worrying that TargetC has not been expanded.
> 
> How so? Note that this PR avoids adding `if (compiling C)` blocks to D semantic routines because this PR is large enough already. And if this PR twists in the wind for months, it'll become very difficult to keep it rebased as the rest of dmd shifts around.

I might be misquoting the spec, but the only guarantee is that `long long >= long >= int >= short >= char`.

So what size shall long be? 8 bytes. What about int? 8 bytes. And short? 8 bytes.  All valid answers.

(One of the targets I regularly test D on has 16-bit ints).

**Comment by WalterBright** [comment]
You're right that long's can be any size. But that is irrelevant to D, as D cannot interface to, say, 36 bit ints.

> (One of the targets I regularly test D on has 16-bit ints).

Digital Mars C still supports 16 bit int targets. But that's irrelevant for D, which only supports interfacing to C with 32 bit ints.

TargetC (and ImportC) only needs to deal with C implementation-defined characteristics that D is capable of interfacing to.

**Comment by WalterBright** [comment]
Tell me more about the 16 bit D target you support.

**Comment by thewilsonator** [comment]
Probably a microcontroller. I remember someone (@luismarques ?) giving a presentation at a dconf (in berlin?) about targeting microcontrollers with D.

**Comment by CrazyPython** [comment]
What's TargetC?

**Comment by ibuclaw** [comment]
> What's TargetC?

Jargon for core contributors only.

**Comment by ibuclaw** [comment]
> You're right that long's can be any size. But that is irrelevant to D, as D cannot interface to, say, 36 bit ints.

Yes, and in those cases a compilation error and halt should occur (sorry, D cannot represent C ints on this target).

However if the C type does match the same size as _a_ D type, then we should use that.  Even if it means that importC `short` is translated as D `long`.

In D code, we hand off the size logic to the druntime library bindings, but if we're being a C compiler, we must deal with it ourselves.

**Comment by luismarques** [comment]
> Probably a microcontroller. I remember someone (@luismarques ?) giving a presentation at a dconf (in berlin?) about targeting microcontrollers with D.

Yeah. I relied on the optimizer to only do 16-bit arithmetic when needed. That's not any different than using C with 8-bit microcontrollers: stuff in C is routinely promoted to int, which is at least 16 bits wide (GCC has a non-standard option to make ints 8 bits wide, but it's highly incompatible and not widely used).

**Comment by ibuclaw** [comment]
> Tell me more about the 16 bit D target you support.

The microcontroller is xstormy16.

It's probably more useful to just paste the macros that are exposed in gcc (ergo, can be used from gdc)

```
#define CHAR_TYPE_SIZE 8
#define BOOL_TYPE_SIZE 8
#define SHORT_TYPE_SIZE 16
#define INT_TYPE_SIZE 16
#define LONG_TYPE_SIZE 32
#define LONG_LONG_TYPE_SIZE 64
#define WCHAR_TYPE_SIZE 32
#define FLOAT_TYPE_SIZE 32
#define DOUBLE_TYPE_SIZE 64
#define LONG_DOUBLE_TYPE_SIZE 64
```
(Snipping out the `_Fract`, `_Accum`, and `_Sat` types).

Other macros that are of interest:
```
#define SIZE_TYPE "unsigned int"
#define PID_TYPE "int"
#define CHAR8_TYPE "unsigned char"
#define CHAR16_TYPE "short unsigned int"
#define CHAR32_TYPE "unsigned int"
#define WCHAR_TYPE "long int"
#define PTRDIFF_TYPE "int"
#define WINT_TYPE "unsigned int"
#define INTMAX_TYPE "long long int"
#define UINTMAX_TYPE "long long unsigned int"
```
(Snipping out stdint.h types, which are optionally set, and I expect not of interest to us).

**Comment by WalterBright** [comment]
@luismarques getting D to interface to 16 bit code is an impressive achievement!

For now, ImportC will focus on the action at the center of the galaxy where all the suns are, rather than way out on an arm :-)

**Comment by WalterBright** [comment]
For some reason, the test coverage results are not included in the display of cparse.d. Is there some configuration thing I need to add?

**Comment by MoonlightSentinel** [comment]
I don't know why the bot doesn't add the comments but you can always refer to their website, see
https://app.codecov.io/gh/dlang/dmd/compare/12507/tree/src/dmd/cparse.d

**Comment by WalterBright** [comment]
Thank you, that's what I need!

**Comment by WalterBright** [comment]
@MoonlightSentinel unfortunately the coverage results for cparse.d are not being updated.

**Comment by MoonlightSentinel** [comment]
All coverage pipelines sucessfully uploaded their report for `cparse.d`, so any problems should be on CodeCov's side.

**Comment by WalterBright** [comment]
At the top the coverage page says:
```
Unmerged Base Commits
The base reference contains commits not present in the head reference.
Please merge the base reference into the head reference to see the most
up to date coverage information.
```
which sounds relevant, but I don't know specifically what it means.

**Comment by WalterBright** [comment]
Probably the most practical way to move forward with coverage testing is to pull this.

**Comment by MoonlightSentinel** [comment]
CodeCov recently changed their algorihtm which determines the base commit ([changelog](https://docs.codecov.io/changelog/release-notes-for-codecov-v4410#manual-parent-picking-override)), seems like it select the wrong base commit for this repository.....

> Probably the most practical way to move forward with coverage testing is to pull this.

You can run the coverage test locally:

```shell
rdmd src/build.d clean
rdmd src/build.d ENABLE_COVERAGE=1
rdmd test/run.d compilable/testcstuff1.d
```

At first glance, a `fail_compilation` test for the various `error`'s is missing.

**Comment by WalterBright** [comment]
I'm going to defer the fail_compilation tests at this time, mainly because they are less important, and due to the fluid nature of this prototype are subject to constant change.

**Comment by WalterBright** [comment]
> You can run the coverage test locally

Thanks, that's working beautifully!

**Comment by WalterBright** [comment]
It's about as far as coverage testing will go without modifying the rest of the front end.

File: src/dmd/dmodule.d
Comment (ibuclaw) [inline src/dmd/dmodule.d:1065]: Could you instead pass a `const` copy/reference to `target.c`, this would allow me to quickly add the other types immediately after this gets pulled.

**Review by atilaneves — APPROVED** [review]

File: src/dmd/dmodule.d
Comment (WalterBright) [reply src/dmd/dmodule.d:1065]: I wanted cparse.d to have minimal dependencies on anything else.

Comment (WalterBright) [reply src/dmd/dmodule.d:1065]: What you could do is move TargetC to globals.d, just the data fields, not the member functions. That would sever its dependencies. Change twchar_t to a size rather than a Type, then the struct can stand alone.

Comment (ibuclaw) [reply src/dmd/dmodule.d:1065]: CParser is already a template though, so the cparse.d module need not import `dmd.target`.

**Comment by MoonlightSentinel** [comment]
So open discussions and remaining issues like https://github.com/dlang/dmd/pull/12507#discussion_r630986639 were ignored again...

File: changelog/ImportC.md
Comment (Geod24) [reply changelog/ImportC.md:90]: Case in point:
- Bindings: https://github.com/bosagora/agora/tree/4154e6eb477b5eab1e6effaa57fb5dda8ad5bc1a/source/scpd/scp
- Source: https://github.com/bosagora/agora/tree/4154e6eb477b5eab1e6effaa57fb5dda8ad5bc1a/source/scpp/src/scp

Same layout, just different folder, everything under `source`. In our case it's C++ code so unaffected, but @MoonlightSentinel 's point is sensible.

File: src/dmd/cparse.d
Comment (dkorpel) [inline src/dmd/cparse.d:439]: I just noticed: this should be `cparseAssignExp`. Now it's actually parsing a D assign expression inherited from `Parser`.

File: src/dmd/cparse.d
Comment (dkorpel) [inline src/dmd/cparse.d:2327]: Same here: `cparseAssignExp`

File: changelog/ImportC.md
Comment (WalterBright) [reply changelog/ImportC.md:90]: > Case in point

There are no .d files in src/scp, so why would existing code put it in the import path for D files?

Comment (WalterBright) [reply changelog/ImportC.md:90]: > The resulting binary will now print 1 instead of 2.

Even if someone did manage to put a .c file with the same name as a .d file in a later directory along the D import path, the odds are about 99.99% it will fail to compile as ImportC won't deal with #include. Even if it did, to not get a compile error the .c file function prototype would have to match the one in the .d one.

Comment (MoonlightSentinel) [reply changelog/ImportC.md:90]: Sure, it's unlikely that such code would compile and exhibit different behaviour. But the real issue here is that this PR introduces a breaking change for existing projects. 
(The silent change in behaviour was just a worst case scenario)

File: src/dmd/cparse.d
Comment (WalterBright) [reply src/dmd/cparse.d:2327]: https://github.com/dlang/dmd/pull/12531

File: src/dmd/cparse.d
Comment (WalterBright) [reply src/dmd/cparse.d:439]: https://github.com/dlang/dmd/pull/12531

**Comment by RUSshy** [comment]
For the preprocessor thing could there be a way to say:


```
@("#DEFINE WHAT EVER")
@("#DEFINE OTHER THING")
import my_header;
```

This will save us from having to put them in a different file

https://github.com/niklas-ourmachinery/zig-themachinery/blob/main/tm_custom_tab.zig#L52

**Comment by ibuclaw** [comment]
This PR caused a regression https://issues.dlang.org/show_bug.cgi?id=23548

**Pull request merged**
